File: app/src/main/java/com/anadi/weatherinfo/repository/InfoLoader.java
package com.anadi.weatherinfo.repository;

import com.anadi.weatherinfo.repository.data.OpenWeatherMapInterface;
import com.anadi.weatherinfo.repository.data.WeatherInfo;

import java.io.IOException;

import okhttp3.ResponseBody;
import retrofit2.Call;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;
import timber.log.Timber;

public class InfoLoader {
    private static final String API_URL = "http://api.openweathermap.org/data/2.5/";
    private static final String API_KEY = "f9dee5683fdf51c7b611df7f57f26926";
    private static final String mUnits = "metric";
    private static String mLang = "en";

    private static final String TEST = API_URL + "weather/?q=Kiev,ua&appid=" + API_KEY + "&units=" + mUnits;

    private static volatile InfoLoader instance;
    private Retrofit mRetrofit;

    // Singleton with double check
    private InfoLoader() {
        mRetrofit = new Retrofit.Builder()
            .baseUrl(API_URL)
            .addConverterFactory(GsonConverterFactory.create())
            .build();
    }

    public static InfoLoader getInstance() {
        InfoLoader result = instance;
        if (result == null) {
            synchronized (InfoLoader.class) {
                result = instance;
                if (result == null) {
                    instance = result = new InfoLoader();
                }
            }
        }
        return result;
    }

    public WeatherInfo load(final String cityName, final Country country) {
        try {
            String location = cityName + "," + country.code.toLowerCase();
            Timber.d("Trying to load weather info for: %s", location);
            Timber.d("TEST = %s", TEST);


            OpenWeatherMapInterface apiService = mRetrofit.create(OpenWeatherMapInterface.class);
            Call<WeatherInfo> call = apiService.getWeather(location, API_KEY, mUnits, mLang);
            Response<WeatherInfo> response = call.execute();

            int statusCode = response.code();
            WeatherInfo weatherInfo = response.body();

            Timber.d("Call (%s) statusCode = %d", location, statusCode);
            // код 200
            if (response.isSuccessful()) {
                Timber.d("Yeah baby!!!");
            }
            else {
                switch(statusCode) {
                    case 404:
                        // страница не найдена. можно использовать ResponseBody, см. ниже
                        break;
                    case 500:
                        // ошибка на сервере. можно использовать ResponseBody, см. ниже
                        break;
                }
                ResponseBody errorBody = response.errorBody();
                if (errorBody != null)
                    Timber.d("Error message: %s", errorBody.string());
            }

            Timber.d("Call (%s) statusCode = %d", location, statusCode);

            if (weatherInfo.weather == null)
                Timber.d("WARNING!!! weatherInfo.weather == null");
            else {
                Timber.d("It's ok somebody created list.");
                Timber.d("weather list size = %d", weatherInfo.weather.size());
            }

            return weatherInfo;
        }
        catch (IOException e) {
            System.err.println(e.getMessage());
            e.printStackTrace();
            return null;
        }
    }
}

File: app/src/main/java/com/anadi/weatherinfo/repository/InfoLoader.kt
package com.anadi.weatherinfo.repository

import com.anadi.weatherinfo.repository.data.OpenWeatherMapInterface
import com.anadi.weatherinfo.repository.data.WeatherInfo
import com.anadi.weatherinfo.utils.WeatherException
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import timber.log.Timber
import java.io.IOException
import kotlin.jvm.Throws

class InfoLoader {
    private val mLang = "en"
    private val mRetrofit: Retrofit

    @Throws(WeatherException::class)
    fun load(cityName: String, country: Country): WeatherInfo {

        val location = cityName + "," + country.code.toLowerCase()

        try {
            Timber.d("Trying to load weather info for: %s", location)
            Timber.d("TEST = %s", TEST)
            val apiService = mRetrofit.create(OpenWeatherMapInterface::class.java)
            val call = apiService.getWeather(location, API_KEY, mUnits, mLang)
            val response = call!!.execute()
            val statusCode = response.code()
            val weatherInfo = response.body() ?: throw WeatherException("response body is empty for $location")

            Timber.d("Call (%s) statusCode = %d", location, statusCode)
            // код 200
            if (response.isSuccessful) {
                Timber.d("Yeah baby!!!")
            } else {
                when (statusCode) {
                    404 -> {
                    }
                    500 -> {
                    }
                }
                val errorBody = response.errorBody()
                if (errorBody != null) {
                    Timber.d("Error message: %s", errorBody.string())
                }
            }
            Timber.d("Call (%s) statusCode = %d", location, statusCode)
            return weatherInfo
        } catch (e: IOException) {
            throw WeatherException("can't load $location", e)
        }
    }

    companion object {
        private const val API_URL = "http://api.openweathermap.org/data/2.5/"
        private const val API_KEY = "f9dee5683fdf51c7b611df7f57f26926"
        private const val mUnits = "metric"
        private const val TEST = API_URL + "weather/?q=Kiev,ua&appid=" + API_KEY + "&units=" + mUnits
    }

    // Singleton with double check
    init {
        mRetrofit = Retrofit.Builder().baseUrl(API_URL).addConverterFactory(GsonConverterFactory.create()).build()
    }
}